### PR TITLE
Experimental MT

### PR DESCRIPTION
**Pull request by ysbaddaden**
Rework of #7214 with much better integration, and acceptable overall performance: -10% slower with `NPROCS=1` and `-Dmt` than with the current scheduler for a simple HTTP::Server stress test using `wrk`.

EDIT: despite achieving good single thread performance despite being thread safe, adding more threads won't improve performance. One of four threads will roughly be equal —allright, 4 threads is slightly faster than 1, but multiple processes are still much faster.

NOTE: I want clean commits for each feature and improvement. I'll force-push revised commits instead of pushing lots of fixups —I already squashed dozens of them.

### CODE REVIEW

**Comment by asterite** [comment]
It would be really cool if you and @bcardiff combined efforts on this topic.

From my point of view, I think it would be cool to have any safe infrastructure for running `spawn` across multiple threads, even if it's slower than now (for example using `GC.enable` and `GC.disable`). Once that's done std code can be upgraded to be multithread-safe, and also work can be done to improve the performance. In the meantime also libraries can be upgraded to consider multithreading and so on. So in a way this would be parallelizing this feature's implementation :-)

**Comment by bararchy** [comment]
Its quite funny , I was 100% that you are both synced on the releases hahah

**Comment by bcardiff** [comment]
Can we get https://github.com/crystal-lang/crystal/pull/7547/commits/2e19060fb007336a9002048297b6d3af9829d8be and https://github.com/crystal-lang/crystal/pull/7547/commits/491bdaea37d91d44049d66b3bc055fd38860fc87 in their own PRs?

I think we can merge those right away.

Note that https://github.com/crystal-lang/crystal/pull/7547/commits/35d30af3e05dc56e6d1d4611b2ebece0ec7368ad is already in master.

File: spec/std/channel_spec.cr [hunk 1]
```diff
@@ -5,6 +5,14 @@ private def yield_to(fiber)
   Crystal::Scheduler.resume(fiber)
 end
 
+private macro it_unless_mt(name, &block)
+  {% if flag?(:mt) %}
+    pending({{name}}) { {{yield}} }
+  {% else %}
+    describe({{name}}) { {{yield}} }
```
Comment (Sija) [inline spec/std/channel_spec.cr]: Why not `it`?

**Comment by ysbaddaden** [comment]
@asterite this is why I pushed the PR instead of keeping it hidden in some branch :)

I'm not aiming to merge for now.

I'm disappointed by the performance of MT for the time being. I'm starting to have acceptable single thread performance, despite being thread safe (overhead is contained) which is a nice achievement, but I don't get any improvement with more threads... channels are awfully slower, and select is disabled —I also broke IO::Evented but that's fixable.

The best performance is currently achieved with current master in "sharding" (distinct processes) and no GC, which has a higher impact on performance than I thought (actual collections, not the safe context switches).

I'll try a Chase and Lev queue for the scheduler. It would avoid a mmap and use a circular queue (that could grow and shrink), and allow to steal many fibers are once, instead of stealing only one at a time.

File: spec/std/channel_spec.cr (hunk 1 above)
Comment (ysbaddaden) [reply spec/std/channel_spec.cr]: Typo :)

**Comment by ysbaddaden** [comment]
@bcardiff extracted and rebased.

**Comment by stakach** [comment]
Not entirely sure if this is the place to discuss this, however I feel like this might be the wrong direction to take the language. Fibers are awesome, they are quite mutation safe, very easy to reason with.

Might it not be better for the language to embrace single thread IO reactor with a thread pool accessible via something like a `ThreadChannel` for processor intensive operations that don't have access to IO operations.

I personally really like the look of #6468 and it feels like the complexity of managing fibers will go up once threads are introduced - given fibers should always run on the same thread.

In short, processes feel like the right way to achieve parallelism however it would be nice to perform some processor intensive operations without blocking the reactor.

**Comment by ysbaddaden** [comment]
#6461 is unrelated, there can be structured concurrency with MT and stealing schedulers.

The complexity you refer to is fearless concurrency which is already a problem, even without threads, it's just not as obvious. This should have it's own discussion. I'll try to open it during the weekend with some ideas (initialize constants at start time, thread safe globals only, *moving* values accross fibers) and links to interesting articles.

As stated above, I'm disappointed with the performance, but I didn't try to add an actor to verify a bcrypt hash, or to calculate a random number in fibers, those cases should be dramatically faster the more threads we add (limited by hardware).

**Comment by ysbaddaden** [comment]
Sigh. Benchmarking is tricky.

With default `wrk` settings (2 threads, 10 connections), running 2 or more threads isn't better than 1 (unless requests are heavily CPU bound) and slower than current master, which is *very* disappointing. But `wrk` with 1 thread and 128 shows something completely different. Using relative results:

```
single-thread 1 process:   × 1.00
single-thread 2 processes: × 1.35
single-thread 3 processes: × 1.65
```
```
multi-thread 1 thread:     × 1.10
multi-thread 2 threads:    × 1.50
multi-thread 3 threads:    × 1.40
```

With enough incoming connections, MT is starting to get faster then running different processes :fireworks:, yet it's still slower with 3+ threads (too much contention? scheduler? gc?).

File: src/concurrent/mt_scheduler.cr [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+      else
+        park
+      end
+    end
+  end
+
+  protected def steal_loop : Fiber?
+    Thread.log "scheduler#steal_loop", self
+
+    100.times do
```
Comment (wontruefree) [inline src/concurrent/mt_scheduler.cr]: I am just curious why 100?
This seems like a high number but I do not do much MT programming.

Comment (ysbaddaden) [reply src/concurrent/mt_scheduler.cr]: Magic number. It should probably be much higher, to avoid parking the thread to quickly, while enqueue should wait for the queue to have some entries before unparking a thread, to make it worth it.

Comment (ysbaddaden) [reply src/concurrent/mt_scheduler.cr]: Maybe it should be adaptive, too, depending on how much time was spent waiting in the previous steal loop, and whether the thread was parked or not.

**Comment by cpunion** [comment]
Great! How to build and test this?

**Comment by asterite** [comment]
I think this can be closed now that #8112 was merged.

@ysbaddaden Sorry this didn't make it through!
